File: 153-find-minimum-in-rotated-sorted-array/find-minimum-in-rotated-sorted-array.cpp
class Solution {
public:
    int findMin(vector<int>& nums) {
        int lo=0, hi=(int)nums.size()-1, mid=-1, z=0;
        while(lo<=hi){
            mid=(lo+hi)/2;
            if(nums[mid]>=nums[lo] && nums[mid]<=nums[hi]){
                if(nums[z]>nums[mid]) z=mid;
                hi=mid-1;
            }
            else if(nums[mid]<=nums[lo] && nums[mid]<=nums[hi]){
                if(nums[z]>nums[mid]) z=mid;
                hi=mid-1;
            }
            else if(nums[mid]>=nums[lo] && nums[mid]>=nums[hi]){
                if(nums[z]>nums[mid]) z=mid;
                lo=mid+1;
            }
            else {
                
            }
        }
        return nums[z];
    }
};

File: 153-find-minimum-in-rotated-sorted-array/find-minimum-in-rotated-sorted-array.java
class Solution {
    public int findMin(int[] nums) {
        int n= nums.length;
        //check edge case
        if(n<=1 || nums[0]<nums[n-1]) return nums[0];

        int low= 0, high= n-1, mid, ans=n-1;
        while(low<=high){
            mid= low+(high-low)/2;
            if(nums[mid]<nums[0]){
                ans= mid;
                high= mid-1;
            }
            else{
                low=mid+1;
            }
        }
        return nums[ans];
    }
}